### PR TITLE
Exclude empty requests in requests list

### PR DESCRIPTION
**Pull request by jtraglia**
Mirrors changes suggested by Felix here:

* https://github.com/ethereum/execution-apis/pull/599
* https://github.com/ethereum/EIPs/pull/8989

### CODE REVIEW

File: specs/electra/validator.md [hunk 1]
```diff
@@ -189,18 +189,38 @@ def prepare_execution_payload(state: BeaconState,
...
+            "value": None,
+            "field": "consolidations",
+            "type": List[ConsolidationRequest, MAX_CONSOLIDATION_REQUESTS_PER_PAYLOAD],
+        },
+    }
+
+    execution_requests = ExecutionRequests()
+    for request in execution_requests_list:
+        request_type, request_data = request[0:1], request[1:]
+        assert request_type in requests, "unexpected request type"
```
Comment (ensi321) [inline specs/electra/validator.md:218]: In Lodestar's implementation, we assert `request_type` to be strictly greater than `previous_request_type` due to the ordering requirement as laid out here: https://github.com/ethereum/execution-apis/blob/5f0c46ecf8df0dd4cecf82c49005159c6ed65417/src/engine/prague.md?plain=1#L82

Implemented check at Lodestar [here](https://github.com/ChainSafe/lodestar/blob/70da4a6f6bdd4323821bda7b78fd0783a9f5bc00/packages/beacon-node/src/execution/engine/types.ts#L478)

Comment (jtraglia) [reply specs/electra/validator.md:218]: Hmm okay good catch. I will update this PR to do the same. Thank you!

File: specs/electra/validator.md [hunk 2]
```diff
@@ -189,18 +189,37 @@ def prepare_execution_payload(state: BeaconState,
 
 *[New in Electra]*
 
-1. The execution payload is obtained from the execution engine as defined above using `payload_id`. The response also includes a `execution_requests` entry containing a list of bytes. Each element on the list corresponds to one SSZ list of requests as defined
-in [EIP-7685](https://eips.ethereum.org/EIPS/eip-7685). The index of each element in the array determines the type of request.
+1. The execution payload is obtained from the execution engine as defined above using `payload_id`. The response also includes a `execution_requests` entry containing a list of bytes. Each element on the list corresponds to one SSZ list of requests as defined in [EIP-7685](https://eips.ethereum.org/EIPS/eip-7685). The first byte of each request is used to determine the request type. Requests must be order by request type in ascending order. As a result, there can only be one instance of each request type per execution requests.
```
Comment (ralexstokes) [inline specs/electra/validator.md:192]:
```suggestion
1. The execution payload is obtained from the execution engine as defined above using `payload_id`. The response also includes a `execution_requests` entry containing a list of bytes. Each element on the list corresponds to one SSZ list of requests as defined in [EIP-7685](https://eips.ethereum.org/EIPS/eip-7685). The first byte of each request is used to determine the request type. Requests must be ordered by request type in ascending order. As a result, there can only be one instance of each request type per execution requests.
```

Comment (jtraglia) [reply specs/electra/validator.md:192]: Ah thanks. My mistake!

File: specs/electra/validator.md [hunk 3]
```diff
@@ -189,18 +189,37 @@ def prepare_execution_payload(state: BeaconState,
...
+    prev_request_type = None
+    execution_requests = ExecutionRequests()
+    for request in execution_requests_list:
+        request_type, request_data = request[0:1], request[1:]
+        assert request_type in requests, "unexpected request type"
+        assert len(request_data) != 0, "empty request data"
+        assert prev_request_type is None or prev_request_type < request_type, "not ascending order"
+        prev_request_type = request_type
+        deserialized_request = ssz_deserialize(requests[request_type]["type"], request_data)
+        setattr(execution_requests, requests[request_type]["field"], deserialized_request)
```
Comment (ralexstokes) [inline specs/electra/validator.md:221]: I do think we should "unroll" this so we can skip `setattr`

something like

```python
if request_type == DEPOSIT_REQUEST_TYPE:
    execution_requests.deposits = ssz_deserialize(List[...], request_data)
elif request_type == WITHDRAWAL_REQUEST_TYPE:
    ...
else:
    ...
```

Comment (ralexstokes) [reply specs/electra/validator.md:221]: then I think `requests` can just be `[DEPOSIT_REQUEST_TYPE, WITHDRAWAL_REQUEST_TYPE, CONSOLIDATION_REQUEST_TYPE]`

**Review by ensi321 — APPROVED** [review]

File: specs/electra/validator.md [hunk 4]
```diff
@@ -189,18 +189,50 @@ def prepare_execution_payload(state: BeaconState,
...
+        WITHDRAWAL_REQUEST_TYPE,
+        CONSOLIDATION_REQUEST_TYPE,
+    ]
+
+    prev_request_type = None
+    for request in execution_requests_list:
+        request_type, request_data = request[0:1], request[1:]
+        assert request_type in request_types, "unexpected request type"
+        assert len(request_data) != 0, "empty request data"
+        assert prev_request_type is None or prev_request_type < request_type, "not ascending order"
```
Comment (ppopth) [inline specs/electra/validator.md:212]:
```suggestion
        assert request_type in request_types
        assert len(request_data) != 0
        assert prev_request_type is None or prev_request_type < request_type
```

We never put the assert error messages, did we? If the statement is not really self-descriptive, I think what we usually do is to put a comment instead?

**Review by ppopth — APPROVED** [review]
Looks good after all

File: specs/electra/validator.md (hunk 4 above)
Comment (jtraglia) [reply specs/electra/validator.md]: I prefer assert messages but yes we generally do comments instead. Fixed.

**Review by hwwhww — APPROVED** [review]

**Review by etan-status — APPROVED** [review]
Cross-checked the block hashes in the test vectors against Nimbus. LGTM ✅

**Pull request merged**